File: app/src/main/java/com/example/todo/view/ActivityMainContainer.java
package com.example.todo.view;

import android.content.Context;
import android.util.AttributeSet;
import android.util.Log;
import android.view.MotionEvent;
import android.widget.FrameLayout;
import android.widget.LinearLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public class ActivityMainContainer extends FrameLayout {
    private static final String TAG = "touch_event";
    private boolean intercepted = false;
    private int lastX, lastY;
    private int firstX;
    private LayoutSwipeListener listener;

    public interface LayoutSwipeListener {
        void swapCursor(int toward);
    }

    public void setSwipeListener(LayoutSwipeListener listener) {
        this.listener = listener;
    }

    public ActivityMainContainer(@NonNull Context context) {
        super(context);
    }

    public ActivityMainContainer(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public ActivityMainContainer(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    public ActivityMainContainer(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
    }

    @Override
    public boolean onInterceptTouchEvent(MotionEvent ev) {
        int x = (int) ev.getX();
        int y = (int) ev.getY();
        switch (ev.getAction()) {
            case MotionEvent.ACTION_DOWN:
                Log.d(TAG, "onInterceptTouchEvent: ACTION_DOWN");
                firstX = x;
                intercepted = false;
                break;
            case MotionEvent.ACTION_MOVE:
                // 如果左右滑动，则将点击事件交由当前view处理；如果上下滑动，则将点击事件交由子级view处理
                int offsetX = x - lastX;
                int offsetY = y - lastY;
                if (offsetX != 0 && offsetY != 0 && Math.abs(offsetX) > Math.abs(offsetY) * 2)
                    intercepted = true;
                Log.d(TAG, "offsetX:" + offsetX + "\toffsetY:" + offsetY);
                Log.d(TAG, "intercepted:" + intercepted);
        }
        lastX = x;
        lastY = y;
        return intercepted;
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        int x = (int) event.getX();
        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                Log.d(TAG, "ACTION_DOWN");
                intercepted = true;
                return true;
            case MotionEvent.ACTION_UP:
                Log.d(TAG, "ACTION_UP");
                int distanceX = x - firstX;
                if (distanceX > 180) listener.swapCursor(-1);
                else if (distanceX < -180) listener.swapCursor(1);
                intercepted = false;
                return true;
        }
        lastX = x;
        return super.onTouchEvent(event);
    }
}

File: app/src/main/java/com/example/todo/view/ActivityMainContainer.kt
package com.example.todo.view

import android.annotation.SuppressLint
import android.content.Context
import android.util.AttributeSet
import android.view.MotionEvent
import android.widget.FrameLayout
import kotlin.math.abs

class ActivityMainContainer @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0,
    defStyleRes: Int = 0
) :
    FrameLayout(context, attrs, defStyleAttr, defStyleRes) {
    private var intercepted: Boolean = false
    private var lastX: Int = 0
    private var lastY: Int = 0
    private var firstX: Int = 0
    private lateinit var listener: LayoutSwipeListener

    interface LayoutSwipeListener {
        fun swapCursor(toward: Int)
    }

    fun setSwipeListener(listener: LayoutSwipeListener) {
        this.listener = listener
    }

    override fun onInterceptTouchEvent(ev: MotionEvent): Boolean {
        val x = ev.x.toInt()
        val y = ev.y.toInt()
        when (ev.action) {
            MotionEvent.ACTION_DOWN -> {
                firstX = x
                intercepted = false
            }
            MotionEvent.ACTION_MOVE -> {
                // 如果是左右滑动事件，则直接拦截
                // 在外部拦截法下，fragment内容不应该实现左右滑动
                val offsetX = x - lastX
                val offsetY = y - lastY
                if (offsetX != 0 && offsetY != 0 && abs(offsetX) > abs(offsetY) * 2)
                    intercepted = true
            }
        }
        lastX = x
        lastY = y
        return intercepted
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(event: MotionEvent): Boolean {
        val x = event.x.toInt()
        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                return true
            }
            MotionEvent.ACTION_UP -> {
                val distance = x - firstX
                if (distance > 180) listener.swapCursor(-1)
                else if (distance < -180) listener.swapCursor(1)
                return true
            }
        }
        return super.onTouchEvent(event)
    }
}